Adding /donoremail endpoint that adds emails to a contact as activity

jordanlaforge opened the pull request:
Emails sent to /donoremail endpoint is added to a contact as previous 
activity history.
Added tests for DonorEmailController and DonorEmail.
If an email is sent that doesn't associate with a contact no emails are 
added.

Resolves: FUN-4744

Commit:
Adding /donoremail endpoint that adds emails to a contact as activity

Emails sent to /donoremail endpoint is added to a contact as previous 
activity history.
Added tests for DonorEmailController and DonorEmail.
If an email is sent that doesn't associate with a contact no emails are 
added.

Resolves: FUN-4744

funmark commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,69 @@
+            frDonorEmails donorEmails = new frDonorEmails(donor);
+            donorEmails.parseEmails(request);
+
+            return donorEmails;
```
why return donorEmails ?

funmark commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,69 @@
+    private String contactId;
+    private frDonor d;
+
+    public frDonorEmails(frDonor d) {
```
why not pass frDonor to parseEmails ?  May not need to store fields

jordanlaforge replied:
Removing constructor and making parseEmails static

Commit:
Adding /donoremail endpoint that adds emails to a contact as activity

Changes based on PR comments.

Resolves: FUN-4744

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,63 @@
+public class frDonorEmails extends frModel {
+    public static final String TYPE = 'DonorEmails';
+
+    public static List<frMapping__c> mappings {
```
I didn't see a way to create these mappings at all with the package installed. Are we intending to let users customize how data about the email flows into salesforce or will we just be populating things with our own defaults?

jordanlaforge replied:
I copied it over because it was needed to extend frModel, but I actually don't need frModel.  I am removing all overrides from frModel.

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,63 @@
+
+            return donor.getContactId();
+        }
+        return '-1';
```
What's the purpose of returning -1 here? should we just return null

jordanlaforge replied:
Changed to null.

Commit:
Removing frModel from DonorEmails and fixing tests.

Commit:
Adding /donoremail endpoint that adds emails to a contact as activity

Changing endpoint to receive one email instead a map of emails.
Updating tests for new change to endpoint

Resolves: FUN-4744

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,41 @@
+            ev.ActivityDateTime = DateTime.newInstance((Long)request.get('sentDate')).dateGMT();
+            ev.WhoId = contactId;
+            ev.Subject = 'Funraise Email - ' + String.valueOf(request.get('subject'));
+            ev.DurationInMinutes = 1;
```
Is this DurationInMinutes field required to create the event?  Seems weird and sorta not applicable to this email

Alexalex1432 replied:
Found the docs for DurationInMinutes:

> If IsAllDayEvent is false, a value must be supplied for either DurationInMinutes or EndDateTime. Supplying values in both fields is allowed if the values add up to the same amount of time.

So looks like instead of populating this value we should populate EndDateTime (which would probably be the same value as the sentDate/ev.ActivityDateTime)

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,41 @@
+    public void parseEmails(Map<String, Object> request, String contactId) {
+        try {
+            System.debug('Creating new event');
+            Event ev = new Event();
```
I thin we need to populate this ev.Type to be Email, right now it looks like a calendar event
https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_event.htm see `Type`

Alexalex1432 replied:
Do you think we should set IsClientManaged to true? Looking at the docs seems like it's built for our use case here

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,41 @@
+            ev.DurationInMinutes = 1;
+            Database.upsert(ev, Event.Fields.fr_Email_ID__c, true);
+        } catch (DMLException e) {
+        	insert new Error__c(Error__c = e.getMessage());
```
Moving forward can we try to add more data to these error__c things?  We most likely have the donor id here so can we add that to the debug message.  Maybe a friendly prefix before adding e.getMessage() like "Failed to insert funraise email for donor id %s. Exception: " +e.getMessage()

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,41 @@
+
+    public void parseEmails(Map<String, Object> request, String contactId) {
+        try {
+            System.debug('Creating new event');
```
I think we'd like to remove these debugs on the final commit for this PR

Commit:
Changing the Events object to an EmailMessage object

Updated tests
Emails sent to salesforce are absorbed as EmailMessages now with an 
EmailMessageRelation object to tie the Email with a Contact

RESOLVES: FUN-4744

Alexalex1432 commented on src/classes/frDonorEmails.cls:
```diff
@@ -0,0 +1,45 @@
+            email.toIds = new String[]{contactId};
+            email.subject = 'Funraise Email - ' + String.valueOf(request.get('subject'));
+            email.MessageDate = DateTime.newInstance((Long)request.get('sentDate')).dateGMT();
+            email.fr_Email_ID__c = String.valueOf(request.get('emailId'));
```
Everything else looks good in this pull request!  Only thing here is we should have this custom field fr_Email_ID__c added as metadata to this repo.  I think that will entail adding it to package.xml and refreshing the metadata from the salesforce server

Commit:
Removing constructor from frDonorEmails, making parseEmails static

Commit:
Updating metadata files for donor emails changes.
